packages/@react-stately/collections/src/useCollection.ts

```diff
@@ -10,21 +10,25 @@
  * governing permissions and limitations under the License.
  */
 
-import {Collection, CollectionBase, Node} from '@react-types/shared';
+import {Collection, CollectionBase, CollectionStateBase, Node} from '@react-types/shared';
 import {CollectionBuilder} from './CollectionBuilder';
 import {useMemo, useRef} from 'react';
 
 type CollectionFactory<T, C extends Collection<Node<T>>> = (node: Iterable<Node<T>>, prev: C | null) => C;
 
-export function useCollection<T extends object, C extends Collection<Node<T>> = Collection<Node<T>>>(props: CollectionBase<T>, factory: CollectionFactory<T, C>, context?: unknown, invalidators: Array<any> = []): C {
+export function useCollection<T extends object, C extends Collection<Node<T>> = Collection<Node<T>>>(props: CollectionStateBase<T, C>, factory: CollectionFactory<T, C>, context?: unknown, invalidators: Array<any> = []): C {
   let builder = useMemo(() => new CollectionBuilder<T>(), []);
 
   let prev = useRef<C>(null);
   return useMemo(() => {
-    let nodes = builder.build(props, context);
+    if (props.collection) {
+      return props.collection;
+    }
+
+    let nodes = builder.build(props as CollectionBase<T>, context);
```

Comment:
So I'm trying to understand the collection changes here. Previously we would use the `items` + render functions or the direct children provided to a collection component (traversing JSX tree) to generate the collection of nodes via CollectionBuilder and the provided factory (aka TableCollection/ListCollection/etc). This meant when `items` changed, we'd get an updated collection, or in Table's case there was some additional invalidationContext stuff that would cause the collection to update its nodes (aka selectionMode, etc)

With this change, we can now provide a collection directly through the props and skip the CollectionBuilder stuff. The initial collection would be built outside via extending [BaseCollection](https://github.com/adobe/react-spectrum/blob/6f343476a38bce2e7cc69cae942e9ec9128bd6bd/packages/react-aria-components/src/Collection.tsx#L334) like shown for the aria Table [component](https://github.com/adobe/react-spectrum/blob/6f343476a38bce2e7cc69cae942e9ec9128bd6bd/packages/react-aria-components/src/Table.tsx#L13). Any updates (like item removals/content updates) are now handled by the new [useCollection](https://github.com/adobe/react-spectrum/blob/6f343476a38bce2e7cc69cae942e9ec9128bd6bd/packages/react-aria-components/src/Collection.tsx#L629) and more specifically by the fake DOM approach mentioned [here](https://github.com/adobe/react-spectrum/blob/aria-components-rfc/rfcs/2023-react-aria-components.md#collections) using [Document](https://github.com/adobe/react-spectrum/blob/aria-component-docs/packages/react-aria-components/src/Collection.tsx#L476) which will return updated collections via .getCollection

Is this accurate? So essentially we won't need the old invalidation context approach since the fake DOM Table would rerender with all the required updates? For example, we used to need to [invalidate](https://github.com/adobe/react-spectrum/blob/9eb87968cf943a94676221729c4b7b8c7e80f26e/packages/%40react-stately/table/src/Row.ts#L67-L72) each Row to rebuild its children to include the checkbox cell if selectionMode went from "none" to "single/multiple", but with the new collections approach the table in the fake DOM would rerender with checkbox cells, giving us a new up to date collection? This last part I'm the most unsure about, still looking at the Table aria components implementation

Reply:
Yep, that's all correct. In the new Table component, the checkbox column isn't a special column that needs to be added while building the collection anymore. It's just a normal column in the table, as you can see in the [docs](https://reactspectrum.blob.core.windows.net/reactspectrum/6f343476a38bce2e7cc69cae942e9ec9128bd6bd/docs/react-aria/Table.html). If you want to provide a component where the checkboxes are automatically added like we do in Spectrum, then you'd create wrapper Row and TableHeader components like [this](https://reactspectrum.blob.core.windows.net/reactspectrum/6f343476a38bce2e7cc69cae942e9ec9128bd6bd/docs/react-aria/Table.html#reusable-wrappers). This adds the extra cells and uses composition to nest the rest of the collection. This all works because the components are just normal React components, and React takes care of rendering everything and efficiently updating our fake DOM instead of needing CollectionBuilder to do that.